File: python/leetcode/010.py
class Solution:
    def isMatch(self, s, p):
        m, n = len(s), len(p)
        dp = [[False for i in range(n + 1)] for j in range(m + 1)]
        dp[0][0] = True
        for i in range(1, n + 1):
            dp[0][i] = i > 1 and '*' == p[i - 1] and dp[0][i - 2]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[j - 1] != '*':
                    dp[i][j] = dp[i - 1][j - 1] and (s[i - 1] == p[j - 1] or '.' == p[j - 1])
                else:
                    dp[i][j] = dp[i][j - 2] or (dp[i - 1][j] and (s[i - 1] == p[j - 2] or p[j - 2] == '.'))
        return dp[m][n]

File: cpp/leetcode/010.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, false));
        dp[0][0] = true;
        for(int i = 1; i <= n; ++i){
            if(i >= 2 && p[i-1] == '*') dp[0][i] = dp[0][i-2];
        }
        for(int i = 1; i <= m; ++i){
            for(int j = 1; j <= n; ++j){
                if(p[j-1] != '*'){
                    dp[i][j] = dp[i-1][j-1] && (p[j-1] == '.' || p[j-1] == s[i-1]);
                } else {
                    dp[i][j] = dp[i][j-2] || dp[i][j-1] || (dp[i-1][j] && (p[j-2] == '.' || p[j-2] == s[i-1]));
                }
            }
        }
        return dp[m][n];
    }
};